Lucene全程加锁导致应用无法正常轮换

### 系统信息

- 外部访问地址: http://localhost:8090/
- 启动时间: 2024-08-28 01:19
- 版本: 2.18.0
- 构建时间: 2024-08-01 10:17
- Git Commit: 58fe872
- Java: IBM Semeru Runtime Open Edition / 21.0.3+9-LTS
- 数据库: PostgreSQL / 16.3 (Debian 16.3-1.pgdg120+1)
- 操作系统: Linux / 6.1.0-23-amd64
- 已激活主题: [Earth 1.4.0](https://github.com/halo-dev/theme-earth.git)
- 已启动插件:
  - [Vditor 编辑器 1.8.0](https://github.com/justice2001/halo-plugin-vditor)
  - [应用市场 1.4.0](https://www.halo.run/store/apps/app-VYJbF)
  - [评论组件 2.4.0](https://github.com/halo-dev/plugin-comment-widget)
  - [RSS 1.3.0](https://github.com/halo-dev/plugin-feed)
  - [搜索组件 1.6.2](https://github.com/halo-dev/plugin-search-widget)
  - [Sitemap 1.1.2](https://github.com/halo-dev/plugin-sitemap)

### 使用的哪种方式运行？

Docker

### 发生了什么？

函数`run.halo.app.search.lucene.LuceneSearchEngine#afterPropertiesSet`在初始化`indexWriter`时，会获取Lucene的write.lock（会先检查是否存在文件，不存在的话会生成这个文件，示为获取锁），由于将`indexWriter`作为成员函数永远保留，因此锁会一直存在，无法多实例部署、无法完成轮换（新实例无法获取锁报错，永远无法通过健康检查）

### 复现步骤

如k8s集群中，重新配置halo版本号导致运行新实例，旧实例在新实例变为“健康”状态后才会退出，而新实例由于无法获取锁，导致永远无法通过健康检查，无法完成轮换。

### 相关日志输出

```shell
__  __      __
   / / / /___ _/ /___
  / /_/ / __ `/ / __ \
 / __  / /_/ / / /_/ /
/_/ /_/\__,_/_/\____/
Version: 2.18.0
2024-08-28T15:16:40.437+08:00  INFO 7 --- [           main] run.halo.app.Application                 : Starting Application v2.18.0 using Java 21.0.3 with PID 7 (/application/BOOT-INF/classes started by unknown in /application)
2024-08-28T15:16:40.439+08:00  INFO 7 --- [           main] run.halo.app.Application                 : No active profile set, falling back to 1 default profile: "default"
2024-08-28T15:16:40.586+08:00  WARN 7 --- [           main] o.s.c.a.AnnotationBeanNameGenerator      : Support for convention-based stereotype names is deprecated and will be removed in a future version of the framework. Please annotate the 'value' attribute in @run.halo.app.theme.finders.Finder with @AliasFor(annotation=Component.class) to declare an explicit alias for @Component's 'value' attribute.
2024-08-28T15:16:41.483+08:00  INFO 7 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Multiple Spring Data modules found, entering strict repository configuration mode
2024-08-28T15:16:41.484+08:00  INFO 7 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data R2DBC repositories in DEFAULT mode.
2024-08-28T15:16:41.600+08:00  INFO 7 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 102 ms. Found 1 R2DBC repository interface.
2024-08-28T15:16:43.591+08:00  INFO 7 --- [           main] org.pf4j.DefaultPluginStatusProvider     : Enabled plugins: []
2024-08-28T15:16:43.592+08:00  INFO 7 --- [           main] org.pf4j.DefaultPluginStatusProvider     : Disabled plugins: []
2024-08-28T15:16:43.594+08:00  INFO 7 --- [           main] org.pf4j.DefaultPluginManager            : PF4J version 3.12.0 in 'deployment' mode
WARNING: A restricted method in java.lang.foreign.Linker has been called
WARNING: java.lang.foreign.Linker::downcallHandle has been called by the unnamed module
WARNING: Use --enable-native-access=ALL-UNNAMED to avoid a warning for this module
2024-08-28T15:16:45.131+08:00  INFO 7 --- [           main] o.a.l.s.MemorySegmentIndexInputProvider  : Using MemorySegmentIndexInput and native madvise support with Java 21 or later; to disable start with -Dorg.apache.lucene.store.MMapDirectory.enableMemorySegments=false
2024-08-28T15:16:45.198+08:00  WARN 7 --- [           main] o.a.l.i.v.VectorizationProvider          : Java vector incubator module is not readable. For optimal vector performance, pass '--add-modules jdk.incubator.vector' to enable Vector API.
2024-08-28T15:16:45.205+08:00  WARN 7 --- [           main] onfigReactiveWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'luceneSearchEngine' defined in class path resource [run/halo/app/config/HaloConfiguration.class]: org.apache.lucene.store.LockObtainFailedException: Lock held by another program: /data/halo/indices/halo/write.lock
2024-08-28T15:16:45.205+08:00  INFO 7 --- [           main] r.h.a.e.c.DefaultControllerManager       : Shutting down 0 controllers...
2024-08-28T15:16:45.206+08:00  INFO 7 --- [           main] r.h.a.e.c.DefaultControllerManager       : Shutdown 0 controllers.
2024-08-28T15:16:45.207+08:00  INFO 7 --- [           main] r.h.a.e.controller.DefaultController     : Disposing controller garbage-collector-controller
2024-08-28T15:16:45.207+08:00  INFO 7 --- [           main] r.h.a.e.controller.DefaultController     : Controller garbage-collector-controller is disposed
2024-08-28T15:16:45.247+08:00  INFO 7 --- [           main] .s.b.a.l.ConditionEvaluationReportLogger : 
Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.
2024-08-28T15:16:45.258+08:00 ERROR 7 --- [           main] o.s.boot.SpringApplication               : Application run failed
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'luceneSearchEngine' defined in class path resource [run/halo/app/config/HaloConfiguration.class]: org.apache.lucene.store.LockObtainFailedException: Lock held by another program: /data/halo/indices/halo/write.lock
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1788) ~[spring-beans-6.1.10.jar:6.1.10]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:600) ~[spring-beans-6.1.10.jar:6.1.10]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522) ~[spring-beans-6.1.10.jar:6.1.10]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337) ~[spring-beans-6.1.10.jar:6.1.10]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.1.10.jar:6.1.10]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335) ~[spring-beans-6.1.10.jar:6.1.10]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200) ~[spring-beans-6.1.10.jar:6.1.10]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:975) ~[spring-beans-6.1.10.jar:6.1.10]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:962) ~[spring-context-6.1.10.jar:6.1.10]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:624) ~[spring-context-6.1.10.jar:6.1.10]
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:66) ~[spring-boot-3.3.1.jar:3.3.1]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754) ~[spring-boot-3.3.1.jar:3.3.1]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456) ~[spring-boot-3.3.1.jar:3.3.1]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:335) ~[spring-boot-3.3.1.jar:3.3.1]
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:149) ~[spring-boot-3.3.1.jar:3.3.1]
	at run.halo.app.Application.main(Application.java:28) ~[classes/:2.18.0]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[na:na]
	at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:91) ~[application/:na]
	at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:53) ~[application/:na]
	at org.springframework.boot.loader.launch.JarLauncher.main(JarLauncher.java:58) ~[application/:na]
Caused by: java.lang.RuntimeException: org.apache.lucene.store.LockObtainFailedException: Lock held by another program: /data/halo/indices/halo/write.lock
	at run.halo.app.search.lucene.LuceneSearchEngine.afterPropertiesSet(LuceneSearchEngine.java:249) ~[classes/:2.18.0]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1835) ~[spring-beans-6.1.10.jar:6.1.10]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1784) ~[spring-beans-6.1.10.jar:6.1.10]
	... 20 common frames omitted
Caused by: org.apache.lucene.store.LockObtainFailedException: Lock held by another program: /data/halo/indices/halo/write.lock
	at org.apache.lucene.store.NativeFSLockFactory.obtainFSLock(NativeFSLockFactory.java:117) ~[lucene-core-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.store.FSLockFactory.obtainLock(FSLockFactory.java:43) ~[lucene-core-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.store.BaseDirectory.obtainLock(BaseDirectory.java:44) ~[lucene-core-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.index.IndexWriter.<init>(IndexWriter.java:955) ~[lucene-core-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at run.halo.app.search.lucene.LuceneSearchEngine.afterPropertiesSet(LuceneSearchEngine.java:245) ~[classes/:2.18.0]
	... 22 common frames omitted
```


### 附加信息

建议为`indexWriter`编写工厂方法，并在不使用时关闭以释放锁。

--- a/application/src/main/java/run/halo/app/search/lucene/LuceneSearchEngine.java
+++ b/application/src/main/java/run/halo/app/search/lucene/LuceneSearchEngine.java
@@ -34,6 +34,7 @@ import org.apache.lucene.document.StoredField;
 import org.apache.lucene.document.StringField;
 import org.apache.lucene.document.TextField;
 import org.apache.lucene.index.DocValuesType;
+import org.apache.lucene.index.IndexNotFoundException;
 import org.apache.lucene.index.IndexWriter;
 import org.apache.lucene.index.IndexWriterConfig;
 import org.apache.lucene.index.Term;
@@ -60,6 +61,7 @@ import org.springframework.core.convert.converter.Converter;
 import org.springframework.lang.NonNull;
 import org.springframework.util.StopWatch;
 import org.springframework.util.StringUtils;
+import reactor.core.Exceptions;
 import run.halo.app.search.HaloDocument;
 import run.halo.app.search.SearchEngine;
 import run.halo.app.search.SearchOption;
@@ -78,9 +80,7 @@ public class LuceneSearchEngine implements SearchEngine, InitializingBean, Dispo
 
     private Analyzer analyzer;
 
-    private IndexWriter indexWriter;
-
-    private SearcherManager searcherManager;
+    private volatile SearcherManager searcherManager;
 
     private Directory directory;
 
@@ -103,12 +103,15 @@ public class LuceneSearchEngine implements SearchEngine, InitializingBean, Dispo
             docs.add(doc);
         });
         var deleteQuery = new TermInSetQuery("id", terms);
-        try {
-            this.indexWriter.updateDocuments(deleteQuery, docs);
-            this.searcherManager.maybeRefreshBlocking();
-            this.indexWriter.commit();
+
+        var writerConfig = new IndexWriterConfig(this.analyzer)
+            .setOpenMode(CREATE_OR_APPEND);
+        try (var indexWriter = new IndexWriter(this.directory, writerConfig)) {
+            indexWriter.updateDocuments(deleteQuery, docs);
         } catch (IOException e) {
-            throw new RuntimeException(e);
+            throw Exceptions.propagate(e);
+        } finally {
+            this.refreshSearcherManager();
         }
     }
 
@@ -117,29 +120,43 @@ public class LuceneSearchEngine implements SearchEngine, InitializingBean, Dispo
         var terms = new LinkedList<BytesRef>();
         haloDocIds.forEach(haloDocId -> terms.add(new BytesRef(haloDocId)));
         var deleteQuery = new TermInSetQuery("id", terms);
-        try {
-            this.indexWriter.deleteDocuments(deleteQuery);
-            this.searcherManager.maybeRefreshBlocking();
-            this.indexWriter.commit();
+        var writerConfig = new IndexWriterConfig(this.analyzer)
+            .setOpenMode(CREATE_OR_APPEND);
+        try (var indexWriter = new IndexWriter(this.directory, writerConfig)) {
+            indexWriter.deleteDocuments(deleteQuery);
         } catch (IOException e) {
-            throw new RuntimeException(e);
+            throw Exceptions.propagate(e);
+        } finally {
+            this.refreshSearcherManager();
         }
     }
 
     @Override
     public void deleteAll() {
-        try {
-            this.indexWriter.deleteAll();
-            this.searcherManager.maybeRefreshBlocking();
-            this.indexWriter.commit();
+        var writerConfig = new IndexWriterConfig(this.analyzer)
+            .setOpenMode(CREATE_OR_APPEND);
+        try (var indexWriter = new IndexWriter(this.directory, writerConfig)) {
+            indexWriter.deleteAll();
         } catch (IOException e) {
-            throw new RuntimeException(e);
+            throw Exceptions.propagate(e);
+        } finally {
+            this.refreshSearcherManager();
         }
     }
 
     @Override
     public SearchResult search(SearchOption option) {
         IndexSearcher searcher = null;
+        var sm = obtainSearcherManager();
+        if (sm.isEmpty()) {
+            // indicate the index is empty
+            var emptyResult = new SearchResult();
+            emptyResult.setKeyword(option.getKeyword());
+            emptyResult.setLimit(option.getLimit());
+            emptyResult.setTotal(0L);
+            emptyResult.setHits(List.of());
+            return emptyResult;
+        }
         try {
             searcher = searcherManager.acquire();
             var queryParser = new StandardQueryParser(analyzer);
@@ -270,28 +287,55 @@ public class LuceneSearchEngine implements SearchEngine, InitializingBean, Dispo
 
     @Override
     public void afterPropertiesSet() throws Exception {
-        try {
-            this.analyzer = CustomAnalyzer.builder()
-                .withTokenizer(StandardTokenizerFactory.class)
-                .addCharFilter(HTMLStripCharFilterFactory.NAME)
-                .addCharFilter(CJKWidthCharFilterFactory.NAME)
-                .addTokenFilter(LowerCaseFilterFactory.NAME)
-                .addTokenFilter(CJKWidthFilterFactory.NAME)
-                .addTokenFilter(CJKBigramFilterFactory.NAME)
-                .build();
-            this.directory = FSDirectory.open(this.indexRootDir);
-            var writerConfig = new IndexWriterConfig(this.analyzer)
-                .setOpenMode(CREATE_OR_APPEND);
-            this.indexWriter = new IndexWriter(this.directory, writerConfig);
-            this.searcherManager = new SearcherManager(this.indexWriter, null);
-            log.info("Initialized lucene search engine");
-        } catch (IOException e) {
-            throw new RuntimeException(e);
+        this.analyzer = CustomAnalyzer.builder()
+            .withTokenizer(StandardTokenizerFactory.class)
+            .addCharFilter(HTMLStripCharFilterFactory.NAME)
+            .addCharFilter(CJKWidthCharFilterFactory.NAME)
+            .addTokenFilter(LowerCaseFilterFactory.NAME)
+            .addTokenFilter(CJKWidthFilterFactory.NAME)
+            .addTokenFilter(CJKBigramFilterFactory.NAME)
+            .build();
+        this.directory = FSDirectory.open(this.indexRootDir);
+        log.info("Initialized lucene search engine");
+    }
+
+    Optional<SearcherManager> obtainSearcherManager() {
+        if (this.searcherManager != null) {
+            return Optional.of(this.searcherManager);
+        }
+        synchronized (this) {
+            // double check
+            if (this.searcherManager != null) {
+                return Optional.of(this.searcherManager);
+            }
+            try {
+                this.searcherManager = new SearcherManager(this.directory, null);
+                return Optional.of(this.searcherManager);
+            } catch (IndexNotFoundException e) {
+                log.warn("Index not ready for creating searcher manager");
+            } catch (IOException e) {
+                log.error("Failed to create searcher manager", e);
+            }
+            return Optional.empty();
         }
     }
 
-    void setIndexWriter(IndexWriter indexWriter) {
-        this.indexWriter = indexWriter;
+    private void refreshSearcherManager() {
+        this.obtainSearcherManager().ifPresent(sm -> {
+            try {
+                sm.maybeRefreshBlocking();
+            } catch (IOException e) {
+                log.warn("Failed to refresh searcher", e);
+            }
+        });
+    }
+
+    Directory getDirectory() {
+        return directory;
+    }
+
+    Analyzer getAnalyzer() {
+        return analyzer;
     }
 
     void setDirectory(Directory directory) {
@@ -323,13 +367,13 @@ public class LuceneSearchEngine implements SearchEngine, InitializingBean, Dispo
         if (this.searcherManager != null) {
             closers.add(this.searcherManager);
         }
-        if (this.indexWriter != null) {
-            closers.add(this.indexWriter);
-        }
         if (this.directory != null) {
             closers.add(this.directory);
         }
         IOUtils.close(closers);
+        this.analyzer = null;
+        this.searcherManager = null;
+        this.directory = null;
         log.info("Destroyed lucene search engine");
     }
 

--- a/application/src/test/java/run/halo/app/search/lucene/LuceneSearchEngineTest.java
+++ b/application/src/test/java/run/halo/app/search/lucene/LuceneSearchEngineTest.java
@@ -1,59 +1,27 @@
 package run.halo.app.search.lucene;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertInstanceOf;
 import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
-import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.ArgumentMatchers.assertArg;
-import static org.mockito.ArgumentMatchers.eq;
-import static org.mockito.Mockito.doThrow;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
 
 import java.io.IOException;
 import java.nio.file.Path;
 import java.time.Instant;
 import java.util.List;
 import java.util.Map;
-import org.apache.lucene.analysis.Analyzer;
-import org.apache.lucene.analysis.standard.StandardAnalyzer;
-import org.apache.lucene.document.Document;
-import org.apache.lucene.index.IndexWriter;
-import org.apache.lucene.index.StoredFields;
-import org.apache.lucene.search.IndexSearcher;
-import org.apache.lucene.search.Query;
-import org.apache.lucene.search.ScoreDoc;
-import org.apache.lucene.search.SearcherManager;
-import org.apache.lucene.search.Sort;
-import org.apache.lucene.search.TopFieldDocs;
-import org.apache.lucene.search.TotalHits;
-import org.apache.lucene.store.Directory;
-import org.assertj.core.util.Streams;
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.store.AlreadyClosedException;
+import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.junit.jupiter.api.io.TempDir;
-import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
 import run.halo.app.search.HaloDocument;
 import run.halo.app.search.SearchOption;
 
 @ExtendWith(MockitoExtension.class)
 class LuceneSearchEngineTest {
-
-    @Mock
-    IndexWriter indexWriter;
-
-    @Mock
-    SearcherManager searcherManager;
-
-    @Mock
-    Directory directory;
-
-    @Mock
-    Analyzer analyzer;
 
     LuceneSearchEngine searchEngine;
 
@@ -62,91 +30,76 @@ class LuceneSearchEngineTest {
 
     @BeforeEach
     void setUp() throws Exception {
-        var searchEngine = new LuceneSearchEngine(tempDir);
-        searchEngine.setIndexWriter(indexWriter);
-        searchEngine.setDirectory(directory);
-        searchEngine.setSearcherManager(searcherManager);
-        searchEngine.setAnalyzer(analyzer);
-        this.searchEngine = searchEngine;
+        this.searchEngine = new LuceneSearchEngine(tempDir);
+        this.searchEngine.afterPropertiesSet();
     }
 
+    @AfterEach
+    void cleanUp() throws Exception {
+        this.searchEngine.destroy();
+    }
 
     @Test
     void shouldAddOrUpdateDocument() throws IOException {
         var haloDoc = createFakeHaloDoc();
         searchEngine.addOrUpdate(List.of(haloDoc));
-        verify(this.indexWriter).updateDocuments(any(Query.class), assertArg(docs -> {
-            var docList = Streams.stream(docs).toList();
-            assertEquals(1, docList.size());
-            var doc = docList.get(0);
-            assertInstanceOf(Document.class, doc);
-            var document = (Document) doc;
-            assertEquals("fake-id", document.get("id"));
-        }));
-        verify(this.searcherManager).maybeRefreshBlocking();
-        verify(this.indexWriter).commit();
+        // validate the index
+        var reader = DirectoryReader.open(searchEngine.getDirectory());
+        assertEquals(1, reader.getDocCount("id"));
     }
 
     @Test
     void shouldDeleteDocument() throws IOException {
+        var haloDoc = createFakeHaloDoc();
+        searchEngine.addOrUpdate(List.of(haloDoc));
+
+        var reader = DirectoryReader.open(searchEngine.getDirectory());
+        assertEquals(1, reader.getDocCount("id"));
+
         this.searchEngine.deleteDocument(List.of("fake-id"));
-        verify(this.indexWriter).deleteDocuments(any(Query.class));
-        verify(this.searcherManager).maybeRefreshBlocking();
-        verify(this.indexWriter).commit();
+
+        reader = DirectoryReader.open(searchEngine.getDirectory());
+        assertEquals(0, reader.getDocCount("id"));
     }
 
     @Test
     void shouldDeleteAll() throws IOException {
+        var haloDoc = createFakeHaloDoc();
+        searchEngine.addOrUpdate(List.of(haloDoc));
+
+        var reader = DirectoryReader.open(searchEngine.getDirectory());
+        assertEquals(1, reader.getDocCount("id"));
+
         this.searchEngine.deleteAll();
 
-        verify(this.indexWriter).deleteAll();
-        verify(this.searcherManager).maybeRefreshBlocking();
-        verify(this.indexWriter).commit();
+        reader = DirectoryReader.open(searchEngine.getDirectory());
+        assertEquals(0, reader.getDocCount("id"));
     }
 
     @Test
     void shouldDestroy() throws Exception {
+        var directory = this.searchEngine.getDirectory();
         this.searchEngine.destroy();
-        verify(this.analyzer).close();
-        verify(this.searcherManager).close();
-        verify(this.indexWriter).close();
-        verify(this.directory).close();
+        assertThrows(AlreadyClosedException.class, () -> DirectoryReader.open(directory));
     }
 
     @Test
-    void shouldAlwaysDestroyAllEvenErrorOccurred() throws Exception {
-        var analyzerCloseError = new IOException("analyzer close error");
-        doThrow(analyzerCloseError).when(this.analyzer).close();
-
-        var directoryCloseError = new IOException("directory close error");
-        doThrow(directoryCloseError).when(this.directory).close();
-        var e = assertThrows(IOException.class, () -> this.searchEngine.destroy());
-        assertEquals(analyzerCloseError, e);
-        assertEquals(directoryCloseError, e.getSuppressed()[0]);
-        verify(this.analyzer).close();
-        verify(this.searcherManager).close();
-        verify(this.indexWriter).close();
-        verify(this.directory).close();
+    void shouldSearchNothingIfIndexNotFound() {
+        var option = new SearchOption();
+        option.setKeyword("fake");
+        option.setLimit(123);
+        option.setHighlightPreTag("<fake-tag>");
+        option.setHighlightPostTag("</fake-tag>");
+        var result = this.searchEngine.search(option);
+        assertEquals(0, result.getTotal());
+        assertEquals("fake", result.getKeyword());
+        assertEquals(123, result.getLimit());
+        assertEquals(0, result.getHits().size());
     }
 
     @Test
-    void shouldSearch() throws IOException {
-        var searcher = mock(IndexSearcher.class);
-        when(this.searcherManager.acquire()).thenReturn(searcher);
-        this.searchEngine.setAnalyzer(new StandardAnalyzer());
-
-        var totalHits = new TotalHits(1234, TotalHits.Relation.EQUAL_TO);
-        var scoreDoc = new ScoreDoc(1, 1.0f);
-
-        var topFieldDocs = new TopFieldDocs(totalHits, new ScoreDoc[] {scoreDoc}, null);
-        when(searcher.search(any(Query.class), eq(123), any(Sort.class)))
-            .thenReturn(topFieldDocs);
-        var storedFields = mock(StoredFields.class);
-
-        var haloDoc = createFakeHaloDoc();
-        var doc = this.searchEngine.getHaloDocumentConverter().convert(haloDoc);
-        when(storedFields.document(1)).thenReturn(doc);
-        when(searcher.storedFields()).thenReturn(storedFields);
+    void shouldSearch() {
+        this.searchEngine.addOrUpdate(List.of(createFakeHaloDoc()));
 
         var option = new SearchOption();
         option.setKeyword("fake");
@@ -154,7 +107,7 @@ class LuceneSearchEngineTest {
         option.setHighlightPreTag("<fake-tag>");
         option.setHighlightPostTag("</fake-tag>");
         var result = this.searchEngine.search(option);
-        assertEquals(1234, result.getTotal());
+        assertEquals(1, result.getTotal());
         assertEquals("fake", result.getKeyword());
         assertEquals(123, result.getLimit());
         assertEquals(1, result.getHits().size());

